KMS v2alpha1 (#35046)
kubernetes/website: content/en/docs/tasks/administer-cluster/kms-provider.md

content_type: task
---
<!-- overview -->
This page shows how to configure a Key Management Service (KMS) provider and plugin to enable secret data encryption. Currently there are two KMS API versions. KMS v1 will continue to work while v2 develops in maturity. If you are not sure which KMS API version to pick, choose v1. 

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

The version of Kubernetes that you need depends on which KMS API version
you have selected. 

- If you selected KMS API v1, any supported Kubernetes version will work fine.

> - If you selected KMS API v2, you should use Kubernetes v{{ skew currentVersion }}

Reviewer: ```suggestion
- If you selected KMS API v2, you should use Kubernetes v{{< skew currentVersion >}}
```
nit: looked at the rendered docs and this isn't formatted: https://deploy-preview-35046--kubernetes-io-vnext-staging.netlify.app/docs/tasks/administer-cluster/kms-provider/

* {{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

{{< tabs name="tab_with_md" >}}
{{% tab name="v1" %}}
* Kubernetes version 1.10.0 or later is required

* etcd v3 or later is required

> 

Reviewer 2: Should a KMS provider be required along with a KMS provider plugin? 
This is stating the obvious but at least it will be clear.

{{< feature-state for_k8s_version="v1.12" state="beta" >}}
{{< /tab >}}
{{% tab name="v2" %}}
* Kubernetes version 1.25.0 or later is required

> 

Reviewer: We need the feature gate set in `kube-apiserver` - `--feature-gates=KMSv2=true` to be able to configure a KMS v2 provider

* {{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

{{< tabs name="tab_with_md" >}}
{{% tab name="v1" %}}
* Kubernetes version 1.10.0 or later is required

* etcd v3 or later is required

{{< feature-state for_k8s_version="v1.12" state="beta" >}}

> {{< /tab >}}

Reviewer 3: Should this be `{{% /tab %}}`?
To be honest, I don't like tabs. The value it brings is very limited and the complexity of the document structure is increased. I don't know how such a page can be converted into a PDF file, for example.

Author: Good catch!

We need this doc to support both v1beta1 and v1alpha1 for awhile. If not tab, do you have other suggestions?

Reviewer 3: The using of tabs is a personal taste. I'd prefer laying out the options one by one, in sub-sections. I can skip sub-sections quickly if it is not relevant to me. This is not a blocker.

Author: updated. PTAL

> * {{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

Reviewer 4: ```suggestion
{{< include "task-tutorial-prereqs.md" >}}

The version of Kubernetes that you need depends on which KMS API version
you have selected. 

- If you selected KMS API v1, any supported Kubernetes version will work fine.
- If you selected KMS API v2, you should use Kubernetes v{{ skew currentVersion }}
  (if you are running a different version of Kubernetes that also supports the v2 KMS
  API, switch to the documentation for that version of Kubernetes).

{{< version-check >}}
```

### KMS v1
* Kubernetes version 1.10.0 or later is required

* etcd v3 or later is required

{{< feature-state for_k8s_version="v1.12" state="beta" >}}

### KMS v2
* Kubernetes version 1.25.0 or later is required

* Set kube-apiserver feature gate: `--feature-gates=KMSv2=true` to configure a KMS v2 provider

> * etcd v3 or later is required

Reviewer 4: ```suggestion
* Your cluster must use etcd v3 or later
```

To configure a KMS provider on the API server, include a provider of type `kms` in the
`providers` array in the encryption configuration file and set the following properties:

> ### KMS v1

Reviewer 5: I do not think it makes sense to separate these by API version since we have the `apiVersion` field.

Author: I think we should separate them because `apiVersion` field is only available after v1.25 and we do want to add it to this doc as that is required to be set to v2 if you are using v2. wdyt?

> ### KMS v1

Reviewer 4: ```suggestion
### KMS v1 {#configuring-the-kms-provider-kms-v1}
```

> * `name`: Display name of the KMS plugin.

Reviewer 5: ```suggestion
* `name`: Display name of the KMS plugin.  Cannot be changed once set.
```

> * `endpoint`: Listen address of the gRPC server (KMS plugin). The endpoint is a UNIX domain socket.

Reviewer 5: Probably worth documenting how to configure the use of an abstract UDS.

Reviewer 2: When discussing with Trousseau's users, often, this is not understood that the socket will be created on the control plane due to the proximity requirements with the api server processes. 
That might be worth mentioning too.

Author: added https://github.com/kubernetes/website/pull/35046/commits/5cdb55d2eb4d9652d46fa1e50a53d712ad8745e7

* `cachesize`: Number of data encryption keys (DEKs) to be cached in the clear.
  When cached, DEKs can be used without another call to the KMS;
  whereas DEKs that are not cached require a call to the KMS to unwrap.
* `timeout`: How long should `kube-apiserver` wait for kms-plugin to respond before
  returning an error (default is 3 seconds).

> ### KMS v2

Reviewer 4: ```suggestion
### KMS v1 {#configuring-the-kms-provider-kms-v2}
```

> * `apiVersion`: API Version for KMS provider (Allowed values: v2, v1 or empty. Any other value will result in an error.)

Reviewer 5: Not sure if this is correct.

Author: https://github.com/kubernetes/website/pull/35046#discussion_r942699082

To configure a KMS provider on the API server, include a provider of type `kms` in the
`providers` array in the encryption configuration file and set the following properties:

{{< tabs name="tab_with_md" >}}
{{% tab name="v1" %}}
* `name`: Display name of the KMS plugin.
* `endpoint`: Listen address of the gRPC server (KMS plugin). The endpoint is a UNIX domain socket.
* `cachesize`: Number of data encryption keys (DEKs) to be cached in the clear.
  When cached, DEKs can be used without another call to the KMS;
  whereas DEKs that are not cached require a call to the KMS to unwrap.
* `timeout`: How long should `kube-apiserver` wait for kms-plugin to respond before
  returning an error (default is 3 seconds).
{{< /tab >}}
{{% tab name="v2" %}}

> * `apiversion`: API Version for KMS provider (Allowed values: `v2`, empty or anything else will default fallback to `v1`)

Reviewer: ```suggestion
* `apiVersion`: API Version for KMS provider (Allowed values: v2, v1 or empty. Any other value will result in an error.)
```

To configure a KMS provider on the API server, include a provider of type `kms` in the
`providers` array in the encryption configuration file and set the following properties:

### KMS v1 {#configuring-the-kms-provider-kms-v1}
* `name`: Display name of the KMS plugin.  Cannot be changed once set.

> * `endpoint`: Listen address of the gRPC server (KMS plugin). The endpoint is a UNIX domain socket.

Author: @enj ptal https://github.com/kubernetes/website/pull/35046/files#diff-01157160365978e29ae4ee6cc3f75f56794f3d2197cb92af5ea304e07ed3cabcR123

To configure a KMS provider on the API server, include a provider of type `kms` in the
`providers` array in the encryption configuration file and set the following properties:

### KMS v1 {#configuring-the-kms-provider-kms-v1}

> * `name`: Display name of the KMS plugin.  Cannot be changed once set.

Reviewer: ```suggestion
* `name`: Display name of the KMS plugin. Cannot be changed once set.
```

* `endpoint`: Listen address of the gRPC server (KMS plugin). The endpoint is a UNIX domain socket.
* `cachesize`: Number of data encryption keys (DEKs) to be cached in the clear.
  When cached, DEKs can be used without another call to the KMS;
  whereas DEKs that are not cached require a call to the KMS to unwrap.
* `timeout`: How long should `kube-apiserver` wait for kms-plugin to respond before
  returning an error (default is 3 seconds).

### KMS v2 {#configuring-the-kms-provider-kms-v2}
* `apiVersion`: API Version for KMS provider (Allowed values: v2, v1 or empty. Any other value will result in an error.)  Must be set to v2 to use the KMS v2 APIs.

> * `name`: Display name of the KMS plugin.  Cannot be changed once set.

Reviewer: ```suggestion
* `name`: Display name of the KMS plugin. Cannot be changed once set.
```

You can develop a KMS plugin gRPC server using a stub file available for Go. For other languages,
you use a proto file to create a stub file that you can use to develop the gRPC server code.

> #### KMS v1

Reviewer 4: ```suggestion
### KMS v1 {#developing-a-kms-plugin-gRPC-server-kms-v1}
```

* Using Go: Use the functions and data structures in the stub file:
  [api.pb.go](https://github.com/kubernetes/kubernetes/blob/release-1.25/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/api.pb.go)
  to develop the gRPC server code 

* Using languages other than Go: Use the protoc compiler with the proto file:
  [api.proto](https://github.com/kubernetes/kubernetes/blob/release-1.25/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/api.proto)
  to generate a stub file for the specific language

> #### KMS v2

Reviewer 4: ```suggestion
### KMS v2 {#developing-a-kms-plugin-gRPC-server-kms-v2}
```

* Using Go: Use the functions and data structures in the stub file:
  [api.pb.go](https://github.com/kubernetes/kubernetes/blob/release-1.25/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v2alpha1/api.pb.go)
  to develop the gRPC server code 

* Using languages other than Go: Use the protoc compiler with the proto file:
  [api.proto](https://github.com/kubernetes/kubernetes/blob/release-1.25/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v2alpha1/api.proto)
  to generate a stub file for the specific language

Then use the functions and data structures in the stub file to develop the server code.

> **Notes:**
> 
> #### KMS v1 Notes

Reviewer 4: ```suggestion
```suggestion
#### Notes

##### KMS v1 {#developing-a-kms-plugin-gRPC-server-notes-kms-v1}
```

* protocol: UNIX domain socket (`unix`)

  The plugin is implemented as a gRPC server that listens at UNIX domain socket. The plugin deployment should create a file on the file system to run the gRPC unix domain socket connection. The API server (gRPC client) is configured with the KMS provider (gRPC server) unix domain socket endpoint in order to communicate with it. 

> #### KMS v2 Notes

Reviewer 4: ```suggestion
##### KMS v2 {#developing-a-kms-plugin-gRPC-server-notes-kms-v2}
```

* protocol: UNIX domain socket (`unix`)

>   The plugin is implemented as a gRPC server that listens at UNIX domain socket. The plugin deployment should create a file on the file system to run the gRPC unix domain socket connection. The API server (gRPC client) is configured with the KMS provider (gRPC server) unix domain socket endpoint in order to communicate with it. Because Linux Abstract sockets do not have concept of ACL (unlike traditional file based sockets), they are subject to Linux networking namespace, so will only be accessible to containers within the same pod (unless host networking is used).

Reviewer 5: ```suggestion
  The plugin is implemented as a gRPC server that listens at UNIX domain socket. The plugin deployment should create a file on the file system to run the gRPC unix domain socket connection. The API server (gRPC client) is configured with the KMS provider (gRPC server) unix domain socket endpoint in order to communicate with it. An abstract Linux socket may be used by starting the endpoint with `/@`, i.e. `unix:///@foo`. Care must be taken when using this type of socket as they do not have concept of ACL (unlike traditional file based sockets). However, they are subject to Linux networking namespace, so will only be accessible to containers within the same pod unless host networking is used.
```

* protocol: UNIX domain socket (`unix`)

  The plugin is implemented as a gRPC server that listens at UNIX domain socket. The plugin deployment should create a file on the file system to run the gRPC unix domain socket connection. The API server (gRPC client) is configured with the KMS provider (gRPC server) unix domain socket endpoint in order to communicate with it. 

##### KMS v2 {#developing-a-kms-plugin-gRPC-server-notes-kms-v2}
* kms plugin version: `v2alpha1`

  In response to procedure call Status, a compatible KMS plugin should return `v2alpha1` as `StatusResponse.Version`, "ok" as `StatusResponse.Healthz` and a keyID (KMS KEK ID) as `StatusResponse.KeyID`

* message version: `v2alpha1`

>   All messages from KMS provider have the version field set to current version v2alpha1.

Reviewer: We no longer have the version field in `EncryptRequest` or `DecryptRequest` for KMS v2, so this can be removed.

* protocol: UNIX domain socket (`unix`)

  The gRPC server should listen at UNIX domain socket.
{{< /tab >}}
{{% tab name="v2" %}}
* kms plugin version: `v2alpha1`

>   In response to procedure call Version, a compatible KMS plugin should return `v2alpha1` as `VersionResponse.version`.

Reviewer: ```suggestion
  In response to procedure call Status, a compatible KMS plugin should return `v2alpha1` as `StatusResponse.Version`, "ok" as `StatusResponse.Healthz` and a keyID (KMS KEK ID) as `StatusResponse.KeyID`
```

ref: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L238-L255

1. `StatusResponse.Healthz` must be ok. Anything other than `ok` will result in health check failure.
2. `StatusResponse.Version` must be `v2alpha1`.
3. `StatusRespose.KeyID` must not be empty. The recommendation is to return the KMS KEK ID as this will be used for rotation.

If either (1), (2) and (3) are not valid, the health check for the KMS plugin will fail.

To encrypt the data:

1. Create a new encryption configuration file using the appropriate properties for the `kms` provider to encrypt resources like Secrets and ConfigMaps:

  **KMS v1**

   ```yaml
   apiVersion: apiserver.config.k8s.io/v1
   kind: EncryptionConfiguration
   resources:
     - resources:
         - secrets
       providers:
         - kms:
             name: myKmsPluginFoo
             endpoint: unix:///tmp/socketfile.sock
             cachesize: 100
             timeout: 3s
         - identity: {}
         - kms:
             name: myKmsPluginBar
             endpoint: unix:///tmp/socketfile.sock
             cachesize: 100
             timeout: 3s

>          - identity: {}

Reviewer 5: Technically one should leave this off if they know all data is stored as encrypted because then having write access to etcd without access to the key vault means that one cannot inject data into the store (i.e. you need either control over the API server config or control over etcd+key vault to do harm).

Reviewer 2: I believe only one identity entry might be present/required and as a last.
Due to the sequential nature of the api server to try each providers, we would first assume data field payload has been encrypted with the main kms, then maybe a former kms from which we want to move away from by doing a replace operation would still be present, and finally the identity provider for pre-kms plugin secrets that we might not have replaced.

Reviewer 5: I understand how the code works, I am pointing out that by recommending the use of the identity provider, we limit the already small value of the encryption at rest support in Kube but allowing an attacker to inject data simply if they have access to etcd.  We should stop doing that because part of the threat model for KMS is not trusting etcd.

To encrypt the data:

1. Create a new encryption configuration file using the appropriate properties for the `kms` provider to encrypt resources like Secrets and ConfigMaps:

>   **KMS v1**

Reviewer 4: I'd use an actual heading here, and set a fragment identifier similar to the other examples. Or you could use tabs here.

         - secrets
       providers:
         - kms:
             name: myKmsPluginFoo
             endpoint: unix:///tmp/socketfile.sock
             cachesize: 100
             timeout: 3s
         - kms:
             name: myKmsPluginBar
             endpoint: unix:///tmp/socketfile.sock
             cachesize: 100
             timeout: 3s
   ```

>   **KMS v2**

Reviewer 4: I'd use an actual heading here, and set a fragment identifier similar to the other examples. Or you could use tabs here.

1. Create a new encryption configuration file using the appropriate properties for the `kms` provider:

> {{< tabs name="tab_with_md" >}}

Reviewer 2: For this section, I would not go with a tab approach especially regarding the nice example in v2 where both can co-exist while the importance if the list order. 
In the current writing, it implies that v1 example can't have v2 in while v2 would tolerate v1 in.

Author: Updated the example for v1, but leaving the tabs since `apiVersion` is new in v2.

> {{% tab name="v1" %}}

Reviewer 2: Could we list the supported resources? 
I had success with ConfigMap as an example.

         - secrets
       providers:
         - kms:
             name: myKmsPluginFoo
             endpoint: unix:///tmp/socketfile.sock
             cachesize: 100
             timeout: 3s
         - kms:
             name: myKmsPluginBar
             endpoint: unix:///tmp/socketfile.sock
             cachesize: 100
             timeout: 3s
   ```

### KMS v2 {#encrypting-your-data-with-the-kms-provider-kms-v2}

   ```yaml
   apiVersion: apiserver.config.k8s.io/v1
   kind: EncryptionConfiguration
   resources:
     - resources:
         - secrets
       providers:
         - kms:
             apiVersion: v2
             name: myKmsPluginFoo
             endpoint: unix:///tmp/socketfile.sock
             cachesize: 100
             timeout: 3s
         - kms:
             name: myKmsPluginBar
             endpoint: unix:///tmp/socketfile.sock
             cachesize: 100
             timeout: 3s
   ```

> 
> For details about the `EncryptionConfiguration` format, please check the

Reviewer 5: ```suggestion
Until the steps defined in `Ensuring all secrets are encrypted` are performed, the `providers` list should end with the `identity: {}` provider to allow unencrypted data to be read.  Once all resources are encrypted, the `identity` provider should be removed to prevent the API server from honoring unencrypted data.

For details about the `EncryptionConfiguration` format, please check the
```

> [API server encryption API reference](/docs/reference/config-api/apiserver-encryption.v1/).

Author: This file is auto-generated and will be updated after v1.25 is out. See https://github.com/kubernetes/website/pull/35046#discussion_r943037151

         - secrets
       providers:
         - kms:

>              apiVersion: v2

Reviewer 5: @aramase is this right?  I thought it would be `v2alpha1`?

Author: https://github.com/kubernetes/kubernetes/pull/111126/files#diff-98f18019b1a1e4502439dc50f54de8977ca2ad2748b1502cea218d21033acf17R12

Reviewer: The allowed values are `v2`, `v1` and default is `v1`
